billy peng ui extract

okwme opened the pull request:
Closes #820 

Description: Replaces Ni components w Tm

okwme commented:
ok so i got voyager to run w the component library by going into `node_modules/vue-loader` and running `yarn`. This shouldn't be necessary of course, but might have something to do with this: 
https://github.com/vuejs/vue-loader/issues/678
also as can be seen in the circle ci test it's a babel issue not understanding "import". I found a similar babel issue here:
https://github.com/babel/babel-loader/issues/149

okwme commented:
wait now this branch it's working as expected after `rm -r node_modules && yarn`
might have been the naming corrections i made on tendermint/ui

circle is still failing of course : \

faboweb commented:
I guess that Babel is not correctly applied to the imported components.

nylira commented:
@faboweb You're right, I just merged a PR to export babel-ified components.

nylira commented:
I spent time trying to fix the tests here and on @tendermint/ui, without much progress. The fact that Babel is transitioning to version 7, Vue is transitioning to version 3, and Jest is falling behind [1] is making this extra difficult. There's a complex dependency chain filled with incompatibilities, and CircleCI throws another wrench into the gears.

An option we can look into in the near future is dropping Babel. We can start using ES modules in Electron already, with things like this [2][3]

[1] https://github.com/facebook/jest/issues/4842
[2] https://gist.github.com/smotaal/f1e6dbb5c0420bfd585874bd29f11c43
[3] https://cli.vuejs.org/guide/browser-compatibility.html#modern-mode

nylira commented:
I'm getting errors like this with `yarn testnet`:

```
         ERROR in ../ui/src/components/TmPageHeader/TmPageHeader.vue
         Module not found: Error: Can't resolve 'vue-loader/node_modules/vue-style-loader' in '/Users/pz/Projects/ui/src/components/TmPageHeader'
          @ ../ui/src/components/TmPageHeader/TmPageHeader.vue 4:2-326
          @ ../ui/src/index.js
          @ ./node_modules/babel-loader/lib!./node_modules/vue-loader/lib/selector.js?type=script&index=0!./app/src/renderer/components/wallet/PageWallet.vue
          @ ./app/src/renderer/components/wallet/PageWallet.vue
          @ ./app/src/renderer/components ^\.\/.*$
          @ ./app/src/renderer/routes.js
          @ ./app/src/renderer/main.js
          @ multi (webpack)-dev-server/client?http://localhost:9080 webpack/hot/dev-server ./app/src/renderer/main.js

         ERROR in ../ui/src/components/TmPart/TmPart.vue
         Module not found: Error: Can't resolve 'vue-loader/node_modules/vue-style-loader' in '/Users/pz/Projects/ui/src/components/TmPart'
          @ ../ui/src/components/TmPart/TmPart.vue 4:2-320
          @ ../ui/src/index.js
          @ ./node_modules/babel-loader/lib!./node_modules/vue-loader/lib/selector.js?type=script&index=0!./app/src/renderer/components/wallet/PageWallet.vue
          @ ./app/src/renderer/components/wallet/PageWallet.vue
          @ ./app/src/renderer/components ^\.\/.*$
          @ ./app/src/renderer/routes.js
          @ ./app/src/renderer/main.js
          @ multi (webpack)-dev-server/client?http://localhost:9080 webpack/hot/dev-server ./app/src/renderer/main.js
```

nylira commented:
I used `module.exports` in the `index.js` of https://github.com/nylira/vue-field and we've been using `vue-field` without problems for a long time. I wonder if we can solve a lot of these issues by not ES6 modules in https://github.com/tendermint/ui/blob/develop/src/index.js

faboweb commented on app/src/renderer/components/common/NiSessionSignIn.vue:
```diff
@@ -118,7 +118,7 @@ export default {
 </script>
 <style lang="stylus">
-.ni-form-group
+.tm-tm-form-group
```
tm-tm?

okwme replied:
nice catch! 😄

faboweb commented on app/src/renderer/components/common/VrToolBar.vue:
```diff
@@ -0,0 +1,31 @@
+<template lang="pug">
```
What is the VrToolBar?

faboweb replied:
Vr = Voyager got it

okwme replied:
🚀

faboweb commented:
This looks pretty decent. Awesome work @okwme and @nylira . What is the status of this PR? What is missing?

okwme commented:
@faboweb still need to improve code coverage, although there's not really any new code besides VrToolBar. Actually maybe removing code damaged the coverage report as the missing governance tests now have a larger effect?

I think we need to figure out a system for deciding what goes into the ui library. There are a lot more common components but they're not being used outside of voyager atm. Should they be included in the library? in the PR? if they stay in voyager should we rename them all to Tm prefix? Vr? I did a search and replace on `ni-` -> `tm-` to fix inherited style changes, not sure if there are other implications that need to be considered.

Also still need to get them published on npmjs.com, i'm not sure how that permission works. @nylira you have access right? will future updates always need to go through you or should we automate it somehow? I hate seeing npm packages that are behind their project repos...

nylira commented:
99% of this PR was due to @okwme awesome efforts. Great job. 

I'd suggest renaming Vr modules to Tm just to keep everything under the same namespace. We can keep single-use Tm components in Voyager for now, but as we continue to build front end projects we'll likely reuse them--and that's when we can move them out from this repo.

If it's simple, I propose removing governance components from the coverage report. We're not working on governance yet. (We will be soon, as first version of governance is coming into the SDK).

I have access to deploying new modules to the tendermint npm organization. I can give you access too, @okwme. But you need to enable 2FA on your NPM account first. Let me know. I don't know if we can automate publishing because every publish requires you to enter your OTP. How do we give our CI the OTP without compromising security?

okwme commented:
* renamed Vr as well as the remaining Ni prefixed components
 * removed governance from the coverage report
 * merged/resolved everything from develop

@nylira I've added 2FA to my npmjs.com account, my username is `okwme`. Happy to publish tendermint/tm and edit this pr to pull directly from the npm repo. 

As for automating w CI I think you're probably right and it's not worth it for the security concern. Maybe instead we just add a git hook to check for the version number and give a reminder to `npm version patch` and `npm publish`?

faboweb commented on app/networks/genesis.json:
```diff
@@ -0,0 +1,1459 @@
+{
```
Why is there this genesis.json now?

okwme replied:
not entirely sure where that came from. I replaced my genesis.json with the one you sent me last week. Does this one not match that one?

faboweb replied:
The location is odd, as genesis.json usually belongs in a folder for a network.

okwme replied:
aha, maybe i had some sloppy drag and drop happening! * removing *

faboweb commented on app/src/renderer/components/common/TmToolBar.vue:
```diff
@@ -0,0 +1,31 @@
+    TmToolBar
+  },
+  mounted() {
+    // console.log(TmToolBar)
```
?

okwme replied:
This used to be called VrToolBar and it extended @tendermint/ui/TmToolBar because all voyager pages have those three slots in common. After the last review from @peng he suggested we name all the components Tm but made the names different withiin the component (tm-tool-bar versus vm-tool-bar) so that they would render without recursion.

faboweb commented on test/unit/specs/components/common/Page404.spec.js:
```diff
@@ -8,13 +9,19 @@ describe("Page404", () => {
+  it("has the expected html structure", async () => {
+    await wrapper.vm.$nextTick()
+    wrapper.update()
+    // console.log(wrapper.vm.$el)
```
?

okwme replied:
$nextTick is added for the scrollbar plugin

faboweb commented on test/unit/specs/components/common/PagePreferences.spec.js:
```diff
@@ -17,7 +17,9 @@ describe("PagePreferences", () => {
 
-  it("has the expected html structure if connected", () => {
+  it("has the expected html structure if connected", async () => {
+    await wrapper.vm.$nextTick()
```
Why do we need this here?

okwme replied:
$nextTick is added for the scrollbar plugin

faboweb commented on test/unit/specs/components/monitor/PageBlock.spec.js:
```diff
@@ -40,7 +40,9 @@ describe("PageBlock", () => {
 
-  it("has the expected html structure", () => {
+  it("has the expected html structure", async () => {
+    await wrapper.vm.$nextTick()
```
Why do we need this here?

okwme replied:
$nextTick is added for the scrollbar plugin

faboweb replied:
Could you maybe add comments describing this?

okwme replied:
In general the timing of component rendering, and the need for a lot of wrapper.updates() occurred after moving the components into an external library that had to be explicitly included into babel during the jest tests.

faboweb commented:
Awesome! :)

okwme commented:
e2e tests failed but pass for me locally 😢
trying again...
